theano/gpuarray/dnn.py

```diff
@@ -1129,6 +1135,8 @@ def dnn_conv3d(img, kerns, border_mode='valid', subsample=(1, 1, 1), dilation=(1
                    shape_i(img, 4, fgraph) - shape_i(kerns, 4, fgraph) + 1)
         out_shp = assert_conv_shape(out_shp)
         out = GpuAllocEmpty(dtype=img.dtype, context_name=ctx_name)(*out_shp)
+        if precision == 'float16':
```

Comment:
Currently, if precission is float16 and we cann dnn_conv() and dnn_conv3d, it will silently upcast it to float32 in some cases. This isn't tested.

I don't like that too much. Why not raise an error in that case?